src/shared/Users.ts

```diff
@@ -0,0 +1,125 @@
+import { DataSourceInstanceSettings } from '@grafana/data';
+import { BackendSrv } from '@grafana/runtime';
+import { queryUntilComplete } from 'core/utils';
+import { QueryResponse } from 'core/types';
+import { User, QueryUsersRequest, QueryUsersResponse } from './types/QueryUsers.types';
+import { QUERY_USERS_MAX_TAKE, QUERY_USERS_REQUEST_PER_SECOND } from './constants/Users.constants';
+
+/**
+ * Represents a utility class for managing and querying user data.
+ */
+export class Users {
+  /**
+   * A cached promise that resolves to an array of users.
+   * This cache is used to avoid redundant user data fetches.
+   */
+  private static _usersCache: Promise<User[]> | null = null;
+
+  /**
+   * A cached promise that resolves to a map of user IDs to their full names.
+   * This cache is used to optimize user name lookups.
+   */
+  private static _usersMapCache: Promise<Map<string, string>> | null = null;
+
+  
+  /**
+   * Retrieves the cached promise for the list of users.
+   * If the cache is not initialized, it triggers the loading of users.
+   */
+  get usersCache(): Promise<User[]> {
+    return this.loadUsers();
+  }
+
+  /**
+   * Retrieves the cached promise for the map of user IDs to full names.
+   * If the cache is not initialized, it generates the map from the user list.
+   */
+  get usersMapCache(): Promise<Map<string, string>> {
+    if (Users._usersMapCache) {
+      return Users._usersMapCache;
+    }
+    return this.usersCache.then(users => {
+      const userMap = new Map<string, string>();
+      users.forEach(user => {
+        const fullName = Users.getUserFullName(user);
+        userMap.set(user.id, fullName);
+      });
+      return userMap;
+    });
+  }
+
+  constructor(readonly instanceSettings: DataSourceInstanceSettings, readonly backendSrv: BackendSrv) {
+    this.loadUsers();
+  }
+
+  /**
+   * Generates the full name of a user by combining their first and last names.
+   * @param user - The user object.
+   * @returns The full name of the user.
+   */
+  public static getUserFullName(user: User): string {
+    return `${user.firstName} ${user.lastName}`;
+  }
+
+  /**
+   * Generates a string containing the user's full name and email address.
+   * @param user - The user object.
+   * @returns A string in the format "FirstName LastName (Email)".
+   */
+  public static getUserNameAndEmail(user: User): string {
+    return `${user?.firstName ?? ''} ${user?.lastName ?? ''} (${user?.email ?? ''})`;
+  }
+
+  /**
+   * Loads the list of users from the backend or returns the cached users if available.
+   * This method ensures that user data is fetched only once and reused across the application.
+   * @returns A promise that resolves to an array of users.
+   */
+  private async loadUsers(): Promise<User[]> {
```

Comment:
I guess I don't understand having both this _and_ `usersCache`. It seems you could just move this implementation into `usersCache`, and delete this method entirely.